src/AdvancedVI.jl

```diff
@@ -38,7 +39,14 @@ Evaluate the value and gradient of a function `f` at `x` using the automatic dif
 - `aux`: Auxiliary input passed to `f`.
 - `out::DiffResults.MutableDiffResult`: Buffer to contain the output gradient and function value.
 """
-function value_and_gradient! end
+function value_and_gradient!(
+    ad::ADTypes.AbstractADType, f, x, aux, out::DiffResults.MutableDiffResult
+)
+    grad_buf = DiffResults.gradient(out)
+    y, _ = DifferentiationInterface.value_and_gradient!(f, grad_buf, ad, x, Constant(aux))
```

Comment:
You would benefit from DI's [preparation mechanism](https://gdalle.github.io/DifferentiationInterface.jl/DifferentiationInterface/stable/explanation/operators/#Preparation), especially for backends like ForwardDiff and ReverseDiff

Reply:
Also for Moooncake.

Reply:
@gdalle @willtebbutt In our case, features to come in the future (like subsampling) will result in `aux.prob` change at every iteration. Is is still valid to re-use the same `prep` object? Or is the idea that I should call `DI.prepare_gradient` every time before calling `DI.value_and_gradient`?

Reply:
Mooncake.jl ought to be fine provided that the type doesn't change. Am I correct in assuming that in the case of subsampling, you're just changing the indices of the data which are subsampled at each iteration?

Reply:
At least for reactant, if a Boolean input changes you need to recompile

Reply:
For enzyme forward mode though, assuming it just contains the one hot vectors, it should be fine to reuse

Reply:
@willtebbutt Yeah, I expect the size and values of specific vectors to change, but not the types.

Reply:
@wsmoses Indeed, my main concern was `ReverseDiff` since I know that compiled tapes won't work. How about `Enzyme` reverse mode?

Reply:
Enzyme reverse mode shouldn’t care since iirc DI does no prep for Enzyme reverse mode.

The size changing will definitely make this invalid for Reactant tho

Reply:
In general, for DI, preparation remains valid if the types and size of inputs and constants don't change. I explain it [here](https://gdalle.github.io/DifferentiationInterface.jl/DifferentiationInterface/stable/explanation/operators/#Reusing-preparation) but it's probably not precise enough, feel free to suggest a better wording.

The main exception is indeed ReverseDiff, as documented [here](https://gdalle.github.io/DifferentiationInterface.jl/DifferentiationInterface/stable/explanation/backends/#ReverseDiff). Preparation creates a tape which depends on the control flow, so you need to make sure that this control flow doesn't change with different inputs / constants or your tape becomes invalid. I know this is not perfect, but since I don't have a way to detect control flow inside a function, the alternative would have been to never use taping. This would be correct 100% of the time but it would make ReverseDiff practically useless.

As for Reactant I don't know, I'm still struggling to make it work.

Reply:
If you need the sizes to change, that is not yet implemented but I have an [open issue](https://github.com/gdalle/DifferentiationInterface.jl/issues/519) and [PR](https://github.com/gdalle/DifferentiationInterface.jl/pull/521) about it that would allow modifying the preparation output `prep`, to account for this size change. The drawback is that you need to do it again if the size changes again, so it's not the same as a one-size-fits-all `prep`.
Can you explain in which context the size changes?

Reply:
@gdalle So what we essentially do is take gradient `f(x, aux)` w.r.t `x`, where we compute stuff over data that is contained inside `aux.prob`. #84 will cause a vector of indices inside `aux.prob` to change, so that we compute `f` over a different subset of data. So the input `x` won't change, but some of the content of `aux` will.

Reply:
If the size of `x` and the size of the contents of `aux.prob` don't change, you should be good to go (except for ReverseDiff as explained above). That would require the minibatches to always have the same size I guess?

Reply:
For most backends, what preparation does is allocate caches that correspond to the various differentiated inputs and contexts. These caches have a specific size which determines whether preparation can be reused, but then inside your function you can create more objects with arbitrary sizes, that's none of my business.

Reply:
>  that's none of my business.

Haha fair enough. 

> (except for ReverseDiff as explained above). That would require the minibatches to always have the same length I guess?

I think compiled tapes actually need everything to be the same except `x` no? But in any case, I think the best course of action would be to expose control on whether to use preparing or not. But I feel that would be a little too complicated for this PR alone.

Reply:
Oh yeah I had forgotten constants. As soon as you use constants then DI recompiles the tape every time anyway

Reply:
Ah I see. Okay, then I guess there's nothing preventing us from using `prepare_gradient`.